File: core/src/jvmMain/java/de/haukesomm/sokoban/core/moving/SimpleMoveAction.java
package de.haukesomm.sokoban.core.moving;

import de.haukesomm.sokoban.core.Direction;
import de.haukesomm.sokoban.core.Entity;
import de.haukesomm.sokoban.core.GameState;

import java.util.ArrayList;

public class SimpleMoveAction implements MoveAction {

    private final Entity entity;
    private final Direction direction;

    public SimpleMoveAction(Entity entity, Direction direction) {
        this.entity = entity;
        this.direction = direction;
    }

    @Override
    public GameState performMove(GameState state) {
        var entities = new ArrayList<>(state.entities());
        entities.remove(entity);

        var movedEntity = new Entity(entity.id(), entity.type(), entity.position().nextInDirection(direction));
        entities.add(movedEntity);

        return new GameState(state.tiles(), entities);
    }
}

File: core/src/commonMain/kotlin/de/haukesomm/sokoban/core/moving/SimpleMoveAction.kt
package de.haukesomm.sokoban.core.moving

import de.haukesomm.sokoban.core.Direction
import de.haukesomm.sokoban.core.Entity
import de.haukesomm.sokoban.core.Position
import de.haukesomm.sokoban.core.state.GameState
import de.haukesomm.sokoban.core.state.transform

class SimpleMoveAction(
    private val position: Position,
    private val direction: Direction
) : MoveAction {

    override fun performMove(state: GameState): GameState =
        state.transform {
            val currentIndex = position.toIndex(state.width)
            val entity = tiles[currentIndex].entity
            tiles[currentIndex] = tiles[currentIndex].copy(entity = null)

            val nextIndex = position.nextInDirection(direction).toIndex(state.width)
            tiles[nextIndex] = tiles[nextIndex].copy(entity = entity)

            moves++
        }
}
